File: ml-benchmark/autolifter/single-pass/mss.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

mss = \xs: List. (fix (\f: List -> {Int, Int}. \xs: List.
  match xs with
    nil _ -> {0, 0}
  | cons {h, t} ->
    let res = f t in
      let new_mps = max 0 (+ h res.2) in
        {max new_mps res.1, new_mps}
  end
) xs).1;

main = single_pass mss;

File: ml-benchmark/autolifter/single-pass/mss.ml
type list_ = Nil | Cons of int * list_

let single_pass v =
  let rec run xs =
    match xs with
    | Nil -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let max a b = if a < b then b else a

let mss xs =
  let rec f xs =
    match xs with
    | Nil -> (0, 0)
    | Cons (h, t) ->
        let res = f t in
        let new_mps =
          match res with
          | (_, b) -> max 0 (h + b)
        in
        let mss_value =
          match res with
          | (a, _) -> max new_mps a
        in
        (mss_value, new_mps)
  in
  match f xs with
  | (a, _) -> a

let program = single_pass mss